src/schema/context.ts

```diff
@@ -162,6 +163,9 @@ export class BIDSContext implements Context {
     this.columns = new ColumnsMap() as Record<string, string[]>
     this.json = {}
     this.associations = {} as Associations
+    if (this.dataset.schema.objects) {  // schema may be empty for some tests
+      datatypeFromDirectory(this.dataset.schema, this)
+    }
```

Comment:
@rwblair This is necessary to allow associations to use `datatype` or `modality` in their selectors. I would like to refactor datatype/modality handling, as that really doesn't need to be part of `filenameIdentify` stages. That will be a separate PR, when I have the time.

I'm realizing now that our existing policy of clearing out some path metadata if file rules fail to match will mean that association selectors get different results from, e.g., sidecar or check selectors. This may need further thinking through.